Make `intro`/`intros` progress on existential variables.

herbelin opened the pull request:
**Kind:** feature

- [X] Corresponding documentation was added / updated.
- [x] Entry added in CHANGES.

It happened a couple of times that I expected `intro` to work on an evar and it did not. A typical example, discussed recently, is:
```coq
Definition d := ltac:(intro x; exact (x*x)). 
(* Ltac call to "intro (ident)" failed. No product even after head-reduction. *)
```
This PR forces the refinement of an evar into a dependent product when `intros` is explicitly called with the `force_flag` (added: which is the default for `intro` and `intros H`).

I wonder what users and other developers think about it.

Added note: this does not affect ssreflect notation `=>`.

Zimmi48 commented:
The usual way of doing this would be with `refine (fun x => _)` but the idea of allowing to use `intro` for this looks reasonable to me.

Zimmi48 commented on CHANGES:
```diff
@@ -13,6 +13,12 @@ Vernacular Commands
 
+Tactics
+
+- Intruduction tactics "intro"/"intros" on a goal which is an
```
typo

SkySkimmer commented:
Maybe add a test that `intros` doesn't create an infinite product. Like
```coq
Definition d' : nat -> _ := ltac:(intros;exact 0).
```

herbelin commented:
@Zimmi48: thanks. @SkySkimmer: added.

herbelin commented:
I added the documentation and it was an opportunity to work with sphinx.

This should probably go in another PR but I moved the few ``:math:`\forall`:g:` ... `` of the tactic chapter into ``:g:`forall ...`` by consistency of style. The spacing is not very smooth with neither of them, but, at least, when in verbatim style, there is a reason for rigid spacing.

What to do with some other inconsistencies I found. E.g., I saw sometimes ``` ``foo`` ``` and sometimes ``` :n:`foo` ```. Is there a reason to do one rather than the other? Which one is best? Are there equivalent? Please excuse me if the information is somewhere and I forgot where. At least, I did not see a reference to `:n:` in the MIGRATING file. By the way, there seems to be cool emacs macro there!

Zimmi48 commented:
> What to do with some other inconsistencies I found. E.g., I saw sometimes ``foo`` and sometimes :n:`foo`. Is there a reason to do one rather than the other? Which one is best? Are there equivalent? Please excuse me if the information is somewhere and I forgot where. At least, I did not see a reference to :n: in the MIGRATING file. By the way, there seems to be cool emacs macro there!

Even after having fixed lots of stuff in the manual, I still can't answer this question, by lack of a doc. I'm going to open an issue about this. Maybe we can hope to get help from @cpitclaudel because @maximedenes seems to have a lot on his plate (VM / native-compiler inconsistencies, GitLab CI, etc.) right now.

Zimmi48 commented:
See #7360.

Zimmi48 commented:
When you refer to a command / tactic, it is preferred to use the special `:cmd:` / `:tacn:` to get links to their documentation.

cpitclaudel commented:
> So, somehow, do I understand correctly that :n: is contents (semantics) and that `` is presentation?

Yes, that's correct :) I'll clarify this when I write things up this week-end.

> So, would it be correct to always prefer :n: for a command/tactic so as to express what we mean, independently of the way it shall eventually be rendered?

If you're describing a notation, use `:n:`; it will render it appropriately.  But if you're referring to a command or tactic documented elsewhere, then use `:cmd:` or `:tacn:`, as they'll link to the relevant docs.

ppedrot commented:
So this PR seems to have been hijacked to discuss documentation principles. Am I right that it is ready to be merge though?

herbelin commented:
 @ppedrot: yes, but I think it would make better sense that I move the 3 independent fixes to the tactics chapter to #7365. I'll do that by tonight if noone objects by then. And then, it is ok.

Zimmi48 commented:
@ppedrot I'll just merge as-is if I were you.

herbelin commented:
The night was apparently long, but promised updates done.

cpitclaudel commented:
Question: does this affect `intros` too? As in, does this mean that if the goal is an evar and I do intros it will now work?

ppedrot commented:
@cpitclaudel It doesn't, otherwise it would loop as explained above. There is an explicit flag to deactivate this behaviour with intros.

cpitclaudel commented:
Super, thanks for clarifying!
